programs/axelar-solana-its/src/processor/interchain_transfer.rs

```diff
@@ -233,6 +233,105 @@ fn build_axelar_interchain_token_execute(
     })
 }
 
+/// Processes a regular interchain transfer initiated by a user account.
+///
+/// This function handles transfers where the source address should be the payer
+/// (user account). It validates that the payer is a user account and not a
+/// program or PDA to ensure proper source attribution in the transfer event.
+pub(crate) fn process_user_interchain_transfer<'a>(
+    accounts: &'a [AccountInfo<'a>],
+    token_id: [u8; 32],
+    destination_chain: String,
+    destination_address: Vec<u8>,
+    amount: u64,
+    gas_value: u64,
+    signing_pda_bump: u8,
+    data: Option<Vec<u8>>,
+) -> ProgramResult {
+    // Check that the payer is a user account, not a program or PDA
+    // We get the payer from the first account
+    let payer = next_account_info(&mut accounts.iter())?;
+
+    // User accounts should be owned by the System Program
+    if payer.owner != &solana_program::system_program::ID {
+        msg!(
+            "Payer is not owned by System Program, owner: {}",
+            payer.owner
+        );
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    if payer.executable {
+        msg!("Payer is executable (program account)");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    process_outbound_transfer(
+        accounts,
+        token_id,
+        destination_chain,
+        destination_address,
+        amount,
+        gas_value,
+        signing_pda_bump,
+        data,
+        *payer.key,
+    )
+}
+
+/// Processes an interchain transfer initiated by a PDA.
+pub(crate) fn process_program_interchain_transfer<'a>(
+    accounts: &'a [AccountInfo<'a>],
+    token_id: [u8; 32],
+    destination_chain: String,
+    destination_address: Vec<u8>,
+    amount: u64,
+    gas_value: u64,
+    signing_pda_bump: u8,
+    source_program_id: Option<Pubkey>,
+    // TODO Seeds
+) -> ProgramResult {
+    let source_id = source_program_id.ok_or_else(|| {
+        msg!("Source program ID is required for program-initiated transfers");
+        ProgramError::InvalidInstructionData
+    })?;
+
+    // The payer should be a PDA owned by the source program
+    let payer = next_account_info(&mut accounts.iter())?;
+    if payer.owner != &source_id {
+        msg!(
+            "Payer account must be owned by the source program. Expected: {}, Got: {}",
+            source_id,
+            payer.owner
+        );
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    if payer.executable {
+        msg!("PDA should not be executable");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    // Check that the account is off the ed25519 curve (indicating it's a PDA)
+    let payer_bytes = payer.key.to_bytes();
+    if solana_program::ed25519_program::check_id(&payer_bytes.into()) {
```

Comment:
I believe `Pubkey` already has a [method](https://docs.rs/solana-sdk/2.3.1/solana_sdk/pubkey/struct.Pubkey.html#method.is_on_curve) out of the box for that

Reply:
Also, I don't think this does what we would expect :thinking: . If I understand this function correctly, it checks if the `payer` == `solana_program::ed25519_program::id()`, isn't it?

Reply:
whaaaat, why: :sob:

Reply:
argh, looking at the source code, it seems it is indeed not available on-chain, hence it panics in the tests (which is actually good)

Reply:
Well, if we get the seeds as input, we'll just rederive the PDA, then this is obsolete I believe

Reply:
Agreed!
Fixed in 7efe1d311660ce2001e3a6b451c24bb34b04b7fe